### PR TITLE
Adds support for Helm and Kustomize for AKS service target

### PR DESCRIPTION
**Pull request by wbreza**
# Helm

This PR adds [Helm](https://helm.sh/) support for using `azd` with AKS.
Resolves: #1618 

> [!IMPORTANT]
> To enable Helm support ensure you have the Helm CLI installed
> Enable the helm feature flag by running `azd config set alpha.aks.helm on`

### What Included so far
- [x] Ability to define a list of helm charts to be installed for each `azd` service
- [x] `azd` service does not require any code/docker files

## `azure.yaml` configuration for `azd` 

- [x] Adds new `helm` configuration section under the `k8s` section
- [x] Ability to define a list of helm repositories / releases to install 

```yaml
name: todo-nodejs-mongo-aks
metadata:
  template: todo-nodejs-mongo-aks@0.0.1-beta
services:
  argocd:
    host: aks
    k8s:
      namespace: argo
      service:
        name: argocd-server
      helm:
        repositories:
          - name: argo
            url: https://argoproj.github.io/argo-helm
        releases:
          - name: argocd
            chart: argo/argo-cd
            verson: 5.51.4
  jupyterhub:
    host: aks
    k8s:
      namespace: jupyterhub
      service:
        name: proxy-public
      helm:
        repositories:
          - name: jupyterhub
            url: https://hub.jupyter.org/helm-chart/
        releases:
          - name: jupyterhub
            chart: jupyterhub/jupyterhub
            version: 3.1.0
```

## `azd deploy` Output

- [x] Adds any referenced helm repositories and/or updates them
- [x] Installs the referenced helm charts
- [x] Waits for helm release to transition to a `deployed` state
- [x] Lists relevant services/ingresses defined within the deployed resources 

image

# Kustomize

> [!IMPORTANT]
> To enable Helm support ensure you have the Kustomize CLI installed
> Enable the helm feature flag by running `azd config set alpha.aks.kustomize on`

Adds support to leverage [kustomize](https://kustomize.io/) as part of k8s deployments.

Includes the following:
- Supports `base` and `variant` configurations
- `edits` that can be run before deployments
- `configMapGenerator` with azd environments

# Configuration

The following new configurations are available within the `azure.yaml` file.

## service.k8s.kustomize

- `dir`: Relative path from the service to your kustomize directory that contains a `kustomization.yaml` file
  - Supports environment variable substitution
- `edits`: Array of edit expression that are applied before deployment
  - Supports environment variable substitution
- `env`: Map of key/value pairs generated before deployment
-   - Map values support environment variable substitution

# Use Cases

The following use cases are supported by this change

## Use Case: Dev want to deploy k8s manifests with basic kustomization

The following configuration will perform a `kubectl apply -k <dir>` command that points to your manifests folder that contains a `kustomization.yaml`.

```yaml
# azure.yaml

name: todo-nodejs-mongo-aks
metadata:
  template: todo-nodejs-mongo-aks@0.0.1-beta
services:
  api:
    project: ./src/api
    language: js
    host: aks
    k8s:
      kustomize:
        dir: ./kustomize/base
```

## Use Case: Dev wants to use overlays to deploy to with different variants

This use case is typically used to have custom configurations for deploying to different stages or environments. 
Examples of this would be `dev`, `test` and `prod`

In this example the user can specify the `${AZURE_ENV_NAME}` environment variable within the kustomize directory to automatically leverage the azd environments as your default overlay convention.

```yaml
# azure.yaml

name: todo-nodejs-mongo-aks
metadata:
  template: todo-nodejs-mongo-aks@0.0.1-beta
services:
  api:
    project: ./src/api
    language: js
    host: aks
    k8s:
      kustomize:
        dir: ./kustomize/overlays/${AZURE_ENV_NAME}
```

## Use Case: Dev wants to modify kustomization.yaml before deployment

A common use case for modifying the kustomization.yaml would be for modifying container [image names/versions](https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/images/) used as part of your deployment

In the following example, specify an `edits` configuration and set any valid `kustomize edit ...` command. Environment variables referenced within the edit command are automatically interpolated from your azd environment variables.

```yaml
# azure.yaml

name: todo-nodejs-mongo-aks
metadata:
  template: todo-nodejs-mongo-aks@0.0.1-beta
services:
  api:
    project: ./src/api
    language: js
    host: aks
    k8s:
      kustomize:
        dir: ./kustomize/overlays/${AZURE_ENV_NAME}
        edits:
          - set image todo-api=${SERVICE_API_IMAGE_NAME}
```

## Use Case: Dev want to use `azd` environment variables within config maps

Config maps or secrets are critical in configuring your k8s clusters.  Since [kustomize does not support any direct environment variable substitution](https://kubectl.docs.kubernetes.io/faq/kustomize/eschewedfeatures/#build-time-side-effects-from-cli-args-or-env-variables) we can leverage the kustomize `configMapGenerator` with a `.env` file.

The `kustomize` configuration section supports a `env` section where one or many key/value pairs can be defined.  This configuration will automatically generate a temporary `.env` file within your kustomization directory that can be used by a `configMapGenerator`

The following configuration will create a `.env` with the specified key/value pairs.

```yaml
# azure.yaml

name: todo-nodejs-mongo-aks
metadata:
  template: todo-nodejs-mongo-aks@0.0.1-beta
services:
  api:
    project: ./src/api
    language: js
    host: aks
    k8s:
      kustomize:
        dir: ./kustomize/overlays/${AZURE_ENV_NAME}
        edits:
          - set image todo-api=${SERVICE_API_IMAGE_NAME}
        env:
          AZURE_AKS_IDENTITY_CLIENT_ID: ${AZURE_AKS_IDENTITY_CLIENT_ID}
          AZURE_KEY_VAULT_ENDPOINT: ${AZURE_KEY_VAULT_ENDPOINT}
          APPLICATIONINSIGHTS_CONNECTION_STRING: ${APPLICATIONINSIGHTS_CONNECTION_STRING}
```

The `configMapGenerator` will now generated a k8s config map with the specified name and contain all the key/value pairs referenced within the `.env` file.

```yaml
# kustomization.yaml

apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

resources:
  - deployment.yaml
  - service.yaml
  - ingress.yaml

configMapGenerator:
  - name: todo-web-config
    envs:
      - .env
```

### CODE REVIEW

File: cli/azd/pkg/project/framework_service.go [hunk 1]
```diff
@@ -26,17 +27,14 @@ const (
 )
 
 func parseServiceLanguage(kind ServiceLanguageKind) (ServiceLanguageKind, error) {
-	if string(kind) == "" {
-		return ServiceLanguageKind(""), fmt.Errorf("language property must not be empty")
-	}
-
```
Comment (vhvb1989) [inline cli/azd/pkg/project/framework_service.go:32]: What happen when language is empty? why is this required on this PR for Helm?

File: cli/azd/pkg/project/framework_service_none.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	"github.com/azure/azure-dev/cli/azd/pkg/async"
+	"github.com/azure/azure-dev/cli/azd/pkg/tools"
+)
+
+type NoneProject struct {
+}
+
+func NewNoneProject() FrameworkService {
+	return &NoneProject{}
+}
```
Comment (vhvb1989) [inline cli/azd/pkg/project/framework_service_none.go:15]: This is confusing.  Are we adding a NoneProject to satisfy azd-current-implementation?  Ideally, implementation details should not drive the design, but the other way around.

Is there any other alternative where we don't need to have such `FrameworkService` (None) ?

I'm guessing this is a hack to enable a service with `host: containerapps` to omit the language and bypass/null the packaging step. 
Can we ignore the frameworkService from the execution-flow when `aca` is the host ? 
It feels easier to see a design/condition for adding/injecting a frameworkService based on the language+host than looking at an frameworkService called None and understanding why it is there.

What happen, for example, if the host is not aca and the language is not defined?  That should still be an error, right?

File: cli/azd/pkg/project/service_target_aks.go [hunk 3]
```diff
@@ -225,6 +264,194 @@ func (t *aksTarget) Deploy(
...
+	if serviceConfig.K8s.Kustomize == nil {
+		return false, nil
+	}
+
+	if !t.featureManager.IsEnabled(featureKustomize) {
+		return false, fmt.Errorf(
+			"Kustomize support is not enabled. Run '%s' to enable it.",
+			alpha.GetEnableCommand(featureKustomize),
+		)
+	}
```
Comment (vhvb1989) [inline cli/azd/pkg/project/service_target_aks.go:324]: Should we just print a warning instead of returning an error?
We could say that `kustomize` configuration detected but the feature is off so it is ignored.  

That would allow folks to turn the feature off and keep running the project w/o getting an error.

Same for Helm

File: cli/azd/pkg/helm/cli.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+	return versionResult.Stdout[1:], nil
+}
+
+// StatusResult is the result of a helm status command
+type StatusResult struct {
+	Name      string     `json:"name"`
+	Info      StatusInfo `json:"info"`
+	Version   float64    `json:"version"`
+	Namespace string     `json:"namespace"`
```
[the remark this answers is not in the document]
Comment (wbreza) [reply cli/azd/pkg/helm/cli.go:151]: They aren't the same since the casing is different and we're primarily using it for unmarshalling so I'm not concerned with adding omitempty at the moment.

File: cli/azd/pkg/project/framework_service.go (hunk 1 above)
Comment (wbreza) [reply cli/azd/pkg/project/framework_service.go:32]: When we talk about using containers we lose the pre-requisite of their having to be any code in the project.  I could use `azd` to provision infra and deploy a helm chart to an AKS cluster without having any of my own code in the project.

In this case there isn't any language and this sets the president.

Moving forward in ACA - users may also want to bring their own container image and again - deploy a prebuilt/public image to ACA without having any local project code.

File: cli/azd/pkg/project/framework_service_none.go (hunk 2 above)
Comment (wbreza) [reply cli/azd/pkg/project/framework_service_none.go:15]: Similar to my comment above - once we start bringing/referencing prebuilt/public containers/charts we lose the requirement of requiring project code that has a language.

Today service targets that `azd` supports that target containers would be ACA and AKS. (and even app service but we don't support this config today)

I did think about not having a `NoneProject` but this then leads us down the path of always checking for the existence of a framework service before calling it.  The `NoneProject` essentially is a `noop` but allows us to more easily work in our existing framework.

I'll think on this a bit more but this is where I initially landed.

File: cli/azd/pkg/project/service_target_aks.go (hunk 3 above)
Comment (wbreza) [reply cli/azd/pkg/project/service_target_aks.go:324]: This is the existing experience precedence we have already set for alpha features in `azd`.  If the feature is disabled, we throw an error and tell them how to enable it and force them to run again.

If we want to change this precedence I'm happy to revisit all the alpha features in a separate PR.

File: cli/azd/pkg/project/framework_service_none.go (hunk 2 above)
Comment (wbreza) [reply cli/azd/pkg/project/framework_service_none.go:15]: Have this `NoneProject` implementation also allows us to keep all the rules for what `none` means in a central location/implementation instead of having to scatter the business rules across multiple components..

Comment (vhvb1989) [reply cli/azd/pkg/project/framework_service_none.go:15]: My preference when it comes to SW is to avoid creating components or abstractions which only makes sense after knowing about how the system works internally but are non-sense by itself.

Here's an example:
a banking system with automatic weekly payment to workers depending on the hours they work.  Imagine there is a component that represents a payment deposit of $0 for the case then the worker didn't work any hour. The system internal implementation requires to create a check and mail it to the worker, so the component ensures the system will just send a $0 check to the worker.  

In such example, the system should not send a check and have a different behavior by design.

So, I see it similar, we are introducing a component to represent something with no meaning by itself just to make internal things to keep working, instead of updating the system to handle the case by design.

So, I would leave this to you and @ellismg to decide.  My suggestion is to make the `host` attribute from a service a `first class attribute` which the system uses to know what operations are allowed and should run or not.

Comment (wbreza) [reply cli/azd/pkg/project/framework_service_none.go]: I just noticed that @ellismg already had a `NoopProject` that satisfied the same requirements. Going to deprecate mine in favor of this.

File: cli/azd/cmd/container.go [hunk 5]
```diff
@@ -480,7 +484,7 @@ func registerCommonDependencies(container *ioc.NestedContainer) {
 
 	// Languages
 	frameworkServiceMap := map[project.ServiceLanguageKind]any{
-		"":                                project.NewDotNetProject,
+		project.ServiceLanguageNone:       project.NewNoneProject,
```
[the remark this answers is not in the document]
Comment (wbreza) [reply cli/azd/cmd/container.go]: I don't know how we want to classify this but an empty language should not default to .NET when targeting container based hosts.

**Review by vhvb1989 — APPROVED** [review]
I wonder if long term we can have both helm and Kustomize automatically pulled by azd (like gh and bicep).  That might be atractive to customers?

**Comment by wbreza** [comment]
> I wonder if long term we can have both helm and Kustomize automatically pulled by azd (like gh and bicep). That might be atractive to customers?

We could, but there is complexity, security and maybe legal implications of doing something like this. My recommendation is that we are consistent and either do this across the board for all our dependencies or none of them.
